Added two new settings to Protection Paladin.

Spike2D opened the pull request:
1. Added Lay on Hands setting to decide at which threshold it will be recommended instead of being a hard set setting.
2. Added Blessing of Spellwarding setting, this way we get recommended this magic immunity when we have big spikes of magical damage.

syrifgit commented on TheWarWithin/Priorities/PaladinProtection.simc:
```diff
@@ -25,10 +25,11 @@ actions.cooldowns+=/fireblood,if=buff.avenging_wrath.remains>8
 actions.mitigation+=/shield_of_the_righteous,if=tanking&buff.shield_of_the_righteous.down&(holy_power.deficit=0|buff.divine_purpose.up)
 actions.mitigation+=/word_of_glory,if=health.pct<wog_health&(holy_power.deficit=0|buff.divine_purpose.up|buff.shining_light_free.up)
 actions.mitigation+=/divine_shield,if=talent.final_stand.enabled&tanking&incoming_damage_5s>ds_damage&!(buff.ardent_defender.up|buff.guardian_of_ancient_kings.up|buff.divine_shield.up|buff.potion.up)
+actions.mitigation+=/blessing_of_spellwarding,if=tanking&talent.blessing_of_spellwarding.enabled&incoming_magic_damage_5s>bosp_damage&!(buff.ardent_defender.up|buff.guardian_of_ancient_kings.up|buff.divine_shield.up|buff.potion.up)
```
`&talent.blessing_of_spellwarding.enabled` isn't necessary, as the talent requirement is baked right into the ability registration, and the entry will be skipped automatically without it.

```LUA
 blessing_of_spellwarding = {
        id = 204018,
        cast = 0,
        -- charges = 1,
        cooldown = function() return ( talent.improved_blessing_of_protection.enabled and 240 or 300 ) * ( 1 - 0.15 * talent.uthers_counsel.rank ) end,
        -- recharge = function() return ( talent.improved_blessing_of_protection.enabled and 240 or 300 ) * ( 1 - 0.15 * talent.uthers_counsel.rank ) end,
        gcd = "spell",
        school = "holy",

        spend = 0.15,
        spendType = "mana",

        talent = "blessing_of_spellwarding",
        ...
        ...
```

syrifgit reviewed (approved):
Looks good, left 1 note about talent checks. Thanks!

Hekili commented on TheWarWithin/PaladinProtection.lua:
```diff
@@ -1922,6 +1922,20 @@ spec:RegisterStateExpr( "wog_health", function ()
 end )
 
 
+spec:RegisterSetting( "loh_health", 65, {
```
65% is a really high default value for something that procs Forbearance, no?

Hekili commented on TheWarWithin/PaladinProtection.lua:
```diff
@@ -1976,6 +1990,23 @@ spec:RegisterStateExpr( "ds_damage", function ()
 end )
 
 
+spec:RegisterSetting( "bosp_damage", 40, {
```
Not sure about this; it seems likely that if you *just took* 40% of your health in magic damage, you probably have already had the damage spike hit and Spellwarding would be wasted.  Are there any logs on WCL that show a big spell damage event followed by more spell damage events in rapid succession?

I'd almost lean toward using the Warrior spell reflect filter or checking for enemies with mana who are casting to trigger BoSP.  @johnnylam88 @syrifgit - what do you guys think?

syrifgit replied:
Sorry I forgot to answer here. I think using the spell reflect solution is probably good. Defensive after taking a big chunk generally makes enough sense for an arbitrary DR, but it's a little weirder with a damage type immunity.

Spike2D replied:
Is the spell list from Warrior updated tho? Also, if we gonna use the Spell Reflect solution we might need to rename some of the stuff, if that's fine.

Spike2D commented:
Implemented the requested changes, and also changed the Paladin BOSP to use a Filter, it will need further refinement on the filter file for big hitting spells that we might suggest using it, but the mechanic is implemented.

Hekili commented:
@johnnylam88 Could you take a look here and give feedback about the addition of the spell filter for Paladins?  I'm wondering if we should make the M+ list and reflect list available via Classes.lua so all specs can use it.

@Spike2D For BoSW, does your implementation account for whether or not you are actively tanking the enemy who is casting?  I don't think we want to recommend burning BoSW (or other mitigation) for a tank buster that isn't the player.

Spike2D commented:
@Hekili at the moment the implementation is literally the same as Spell Reflect on the Warrior, I could potentially make it so it only recommends it on Prot Paladin, as at the moment its only if it's targeting you

johnnylam88 commented:
We can easily move the reflect list into `TheWarWithin/Classes.lua` as it's just a bunch of spells broken down by zone, NPC, and spell ID. Instead of just storing the name in the table, we could make it into a table with additional fields, e.g.:
```
[ 123456 ] = {
    name = "Bad Juju",
    spell_reflection = true,
    antimagic_shell = true,
    blessing_of_spellwarding = true,
}
```
Then within each spec's Lua file, you could build the table at load-time from the larger database to only have what you're interested in.

Spike2D commented:
I think that's a lovely solution @johnnylam88, that way we can just toggle on and off as we need it per season or whatever.

johnnylam88 commented:
I can work up a PR with the changes for warrior, and you could adjust for prot pally after it's done. It should be quick :-)

Spike2D commented:
Perfect with me, once the PR is done and up I'll adjust it and update this one, @Hekili is that good with you too?

johnnylam88 commented:
I've got the databases coalesced. At addon load time, I generate the interrupt and spell reflection tables, if needed, for the class, and I use those lookup tables in the abilities. I'll file PR later today with the full changes.

johnnylam88 commented:
With the above changes, the way I would implement suggesting Blessing of Spellwarding for Protection Paladin is to just create a `spellwardingFilters` table in `TheWarWithin/PaladinProtection.lua` in the same way that you see in `TheWarWithin\Warrior\ReflectableSpells.lua`, and then also go through the main database in `TheWarWithin/Classes.lua` and just add `blessing_of_spellwarding = true,` in the appropriate places.

Spike2D commented:
Done, files have been updated to use the new system and spells added that are relevant to BoSP

Hekili commented:
Please fix your indentation and ensure out-of-scope changes are removed (i.e., your PR will revert recent fixes to Consecration, etc.).  It does look good, in terms of using the same infrastructure that @johnnylam88 set up.
